### PR TITLE
Remove PyYAML dependency from Python bindings.

### PR DESCRIPTION
**Pull request by ScottTodd**
Keeping the list of dependencies small will make it easier to depend on us. Installing `iree-compiler` and `iree-runtime` brings these packages along:
```
Installing collected packages: mpmath, sympy, PyYAML, numpy, iree-runtime, iree-compiler
Successfully installed PyYAML-6.0.2 iree-compiler-20240816.987 iree-runtime-20240816.987 mpmath-1.3.0 numpy-2.0.1 sympy-1.13.2
```

Unless LLVM/MLIR somehow needs PyYAML (e.g. for linalg opdsl generators / tests), I think we can drop it now. `git log -S pyyaml` points to only these commits: https://github.com/iree-org/iree/commit/dd787a74204360592d8f2600ba1a3bf4f0f52ebc, https://github.com/iree-org/iree/commit/3149b6d191da1ac55c5ed179337a3c5fb58dbdcd.

* PyYAML is still used by [`build_tools/github_actions/configure_ci.py`](https://github.com/iree-org/iree/blob/main/build_tools/github_actions/configure_ci.py) and [`build_tools/github_actions/configure_ci_test.py`](https://github.com/iree-org/iree/blob/main/build_tools/github_actions/configure_ci_test.py). I just turned off that test in CI. We could install PyYAML (via a `test_dependencies.txt`) to keep running it, but I think we should focus on just what the core project truly needs.
* `build_tools/testing/gen_test_matrix.py` hasn't been used in a while and that sort of test generator should now live in an external repository like https://github.com/iree-org/iree-test-suites

### CODE REVIEW

File: runtime/bindings/python/iree/runtime/build_requirements.txt [hunk 1]
```diff
@@ -14,6 +14,3 @@ sympy==1.12.1
 # TODO: nanobind is used in the runtime but the compiler uses pybind and
 # removing this breaks CI bots; remove this.
 pybind11>=2.10.1
-
-# TODO: not required but needed for weird CI reasons; remove this.
-PyYAML>=5.4.1
```
Comment (ScottTodd) [inline runtime/bindings/python/iree/runtime/build_requirements.txt:19]: https://github.com/iree-org/iree/actions/runs/10426541633/job/28879661057?pr=18262#step:5:3110
```
1530/1568 Test #2224: iree/build_tools/github_actions/configure_ci_test ............................................................***Failed    0.08 sec
Traceback (most recent call last):
  File "/work/build_tools/github_actions/configure_ci_test.py", line 12, in <module>
    import configure_ci
  File "/work/build_tools/github_actions/configure_ci.py", line 50, in <module>
    import yaml
ModuleNotFoundError: No module named 'yaml'
```

lol. Time for a `test_requirements.txt`.

File: compiler/setup.py [hunk 2]
```diff
@@ -463,7 +463,6 @@ def find_git_submodule_revision(submodule_path):
     },
     install_requires=[
         "numpy",
-        "PyYAML",
         "sympy",
```
Comment (ScottTodd) [inline compiler/setup.py:466]: This `sympy` dep appears to be coming from torch-mlir's fx_importer. I left a comment where it was introduced: https://github.com/iree-org/iree/pull/17662#discussion_r1722496958. I think we could move it down into `extras_require`.

Comment (marbre) [reply compiler/setup.py:466]: Yeah, sympy is imported at https://github.com/llvm/torch-mlir/blob/7f886cc2703cb3deec6e95fb10b4aca00ef70da6/test/python/fx_importer/sympy_to_affine_expr_test.py#L9. Worth to try if we can move it into `extras_require`.

**Review by marbre — APPROVED** [review]
Thanks for pushing down the number of deps 😃

**Pull request merged**